### DISCUSSION TITLE
Right approach for creating a migration to alter a table by adding a not-null column

### DISCUSSION THREAD

**Discussion by jiggneshhgohel** [jeremyevans/sequel#2070] [General]
I am using rom-rb which internally uses Sequel and my DB is Postgres. Following are the gem versions being used.

```
gem 'rom', '~> 5.3'
gem 'rom-sql', '~> 3.6', '>= 3.6.1'

gem "pg"
```

I created following DB migration to alter an existing table by adding a not-null column with a default value:

```
  ROM::SQL.migration do
    up do
      alter_table(:my_table_name) do
        add_foreign_key :my_new_column, :other_table_name
        add_index :my_new_column
        set_column_default :my_new_column, 0
        set_column_not_null :my_new_column
      end
    end

    down do
      alter_table(:my_table_name) do
        drop_column :my_new_column
      end
    end
  end
```

But running that migration causes following error:

```
Caused by:
PG::NotNullViolation: ERROR:  column "my_new_column" contains null values
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:171:in `exec'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:171:in `block in execute_query'
.../gems/sequel-5.67.0/lib/sequel/database/logging.rb:38:in `log_connection_yield'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:171:in `execute_query'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:159:in `block in execute'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:136:in `check_disconnect_errors'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:159:in `execute'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:532:in `_execute'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:348:in `block (2 levels) in execute'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:555:in `check_database_errors'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:348:in `block in execute'
.../gems/sequel-5.67.0/lib/sequel/connection_pool/threaded.rb:88:in `hold'
.../gems/sequel-5.67.0/lib/sequel/database/connecting.rb:293:in `synchronize'
.../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:348:in `execute'
.../gems/sequel-5.67.0/lib/sequel/database/query.rb:50:in `execute_dui'
.../gems/sequel-5.67.0/lib/sequel/database/query.rb:43:in `execute_ddl'
.../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:436:in `block in apply_alter_table'
.../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:436:in `each'
.../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:436:in `apply_alter_table'
.../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:449:in `apply_alter_table_generator'
.../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:70:in `alter_table'
db/migrate/20230812085653_my_migration.rb:13:in `block (2 levels) in <top (required)>'
.../gems/sequel-5.67.0/lib/sequel/extensions/migration.rb:110:in `instance_exec'
.../gems/sequel-5.67.0/lib/sequel/extensions/migration.rb:110:in `apply'
.../gems/sequel-5.67.0/lib/sequel/extensions/migration.rb:695:in `block (2 levels) in run'
.../gems/sequel-5.67.0/lib/sequel/database/transactions.rb:258:in `_transaction'
.../gems/sequel-5.67.0/lib/sequel/database/transactions.rb:233:in `block in transaction'
.../gems/sequel-5.67.0/lib/sequel/connection_pool/threaded.rb:92:in `hold'
.../gems/sequel-5.67.0/lib/sequel/database/connecting.rb:293:in `synchronize'
.../gems/sequel-5.67.0/lib/sequel/database/transactions.rb:195:in `transaction'
.../gems/sequel-5.67.0/lib/sequel/extensions/migration.rb:482:in `checked_transaction'
.../gems/sequel-5.67.0/lib/sequel/extensions/migration.rb:694:in `block in run'
.../gems/sequel-5.67.0/lib/sequel/extensions/migration.rb:691:in `each'
.../gems/sequel-5.67.0/lib/sequel/extensions/migration.rb:691:in `run'
.../gems/sequel-5.67.0/lib/sequel/extensions/migration.rb:409:in `run'
.../gems/rom-sql-3.6.1/lib/rom/sql/migration/migrator.rb:31:in `run'
.../gems/rom-sql-3.6.1/lib/rom/sql/migration.rb:139:in `block in run_migrations'
.../gems/rom-sql-3.6.1/lib/rom/sql/migration.rb:77:in `with_gateway'
.../gems/rom-sql-3.6.1/lib/rom/sql/migration.rb:138:in `run_migrations'
.../gems/rom-sql-3.6.1/lib/rom/sql/tasks/migration_tasks.rake:13:in `run_migrations'
.../gems/rom-sql-3.6.1/lib/rom/sql/tasks/migration_tasks.rake:72:in `block (2 levels) in <top (required)>'
.../gems/bundler-2.4.12/lib/bundler/cli/exec.rb:58:in `load'
.../gems/bundler-2.4.12/lib/bundler/cli/exec.rb:58:in `kernel_load'
.../gems/bundler-2.4.12/lib/bundler/cli/exec.rb:23:in `run'
.../gems/bundler-2.4.12/lib/bundler/cli.rb:492:in `exec'
.../gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
.../gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
.../gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
.../gems/bundler-2.4.12/lib/bundler/cli.rb:34:in `dispatch'
.../gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
.../gems/bundler-2.4.12/lib/bundler/cli.rb:28:in `start'
.../gems/bundler-2.4.12/exe/bundle:45:in `block in <top (required)>'
.../gems/bundler-2.4.12/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
.../gems/bundler-2.4.12/exe/bundle:33:in `<top (required)>'
.../bin/bundle:25:in `load'
.../bin/bundle:25:in `<main>'
.../bin/ruby_executable_hooks:22:in `eval'
.../bin/ruby_executable_hooks:22:in `<main>'
```

Initially I thought may be the default 0 value for a foreign key is causing problem but even after changing it to a valid value the error persisted. So to get past the error I took following approach:

1. Created a migration which adds an optional column `my_new_column` to `my_table_name`  (please see Migration-1 ahead).
2. Manually update `my_table_name`'s  `my_new_column` to correct values.
3. Created another migration which makes the `my_table_name`'s  `my_new_column` not-null (please see Migration-2 ahead)..

**Migration-1**

```
ROM::SQL.migration do
  up do
    alter_table(:my_table_name) do
      add_foreign_key :my_new_column, :other_table_name
      add_index :my_new_column
      set_column_allow_null :my_new_column
    end
  end

  down do
    alter_table(:my_table_name) do
      drop_column :my_new_column
    end
  end
end
```

**Migration-2**

```
ROM::SQL.migration do
  up do
    alter_table(:my_table_name) do
      set_column_default :my_new_column, 0
      set_column_not_null :my_new_column
    end
  end

  down do
    alter_table(:my_table_name) do
      set_column_default :my_new_column, nil
      set_column_allow_null :my_new_column
    end
  end
end
```

But somehow I am not finding this multiple migrations approach correct and esp running the migrations on different development envs already containing existing tables can become annoyance. So that makes me ask the question what is the right approach in Sequel to create a migration for altering an existing table by adding a not-null column with a default value?

Also a subjective thought which came to my mind that when using `set_column_not_null :my_new_column` shouldn't it support option for specifying a default value? I mean without a default value adding a not-null column to an existing table doesn't make sense.

---------------------------------

There is one more weird issue I encountered related to a migration which alters a table by adding a null column. Following is a `up` part of the migration:

  ```
    up do
      alter_table(:my_table) do
        add_foreign_key :my_new_column, :other_table_name
        set_column_allow_null :my_new_column
      end
    end
  ```

When ran the migration I encountered following error:

  ```
  Caused by:
    PG::UndefinedColumn: ERROR:  column "my_new_column" of relation "my_table" does not exist
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:171:in `exec'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:171:in `block in execute_query'
    .../gems/sequel-5.67.0/lib/sequel/database/logging.rb:38:in `log_connection_yield'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:171:in `execute_query'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:159:in `block in execute'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:136:in `check_disconnect_errors'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:159:in `execute'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:532:in `_execute'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:348:in `block (2 levels) in execute'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:555:in `check_database_errors'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:348:in `block in execute'
    .../gems/sequel-5.67.0/lib/sequel/connection_pool/threaded.rb:88:in `hold'
    .../gems/sequel-5.67.0/lib/sequel/database/connecting.rb:293:in `synchronize'
    .../gems/sequel-5.67.0/lib/sequel/adapters/postgres.rb:348:in `execute'
    .../gems/sequel-5.67.0/lib/sequel/database/query.rb:50:in `execute_dui'
    .../gems/sequel-5.67.0/lib/sequel/database/query.rb:43:in `execute_ddl'
    .../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:436:in `block in apply_alter_table'
    .../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:436:in `each'
    .../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:436:in `apply_alter_table'
    .../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:449:in `apply_alter_table_generator'
    .../gems/sequel-5.67.0/lib/sequel/database/schema_methods.rb:70:in `alter_table'
    db/migrate/20230807094006_my_migration.rb:5:in `block (2 levels) in <top (required)>'  
  ```

 However when I modified the `up` part in following manner the migration executed successfully:

  ```
    up do
      alter_table(:my_table) do
        add_foreign_key :my_new_column, :other_table_name
       add_index :my_new_column #<-------- Added this
        set_column_allow_null :my_new_column
      end
    end
  ```

I found that behaviour weird.

Please share thoughts on the correct approaches to be employed if I am wrong anywhere in my usage of the Migration API.

Thanks.

**Comment by rgalanakis** [comment]
A lot of this stems from thinking there is a lot more magic going in Sequel. Sequel is mostly mapping SQL concepts to Ruby, as directly as possible without breaking either idiom. If you see an `add_column` method, it adds a column. If you see a `set_default` method, it's setting a default. Calling `add_column` and then `set_default` doesn't try to be clever and combine the default into the create- it treats them as separate.

The create column, set not null, set default are treated as three operations.

You can *create* a column as not-null-with-a-default, ie `ALTER TABLE x ADD COLUMN y TEXT NOT NULL DEFAULT ''`, but there is no way to alter all these things on a column at once.

This is the source of some of your issues above:

```
        add_foreign_key :my_new_column, :other_table_name
        add_index :my_new_column
        set_column_default :my_new_column, 0
        set_column_not_null :my_new_column
```

You are adding a column, then altering it twice. You should set the default and not-null as part of the create, ie `add_foreign_key :my_new_column, :other_table_name, default: 0, null: false`. However the 0 is probably not a valid ID.

> I mean without a default value adding a not-null column to an existing table doesn't make sense.

What's the default value of a user's required email address, or the default value of a `value` column in a `phone_numbers` table? There are endless cases where there is no logical default, or the logical default should be left up to the application (like the initial state in a state machine).

> There is one more weird issue I encountered related to a migration which alters a table by adding a null column.

Jeremy would need to confirm but this may be a bug/edge case. You can/should use `add_foreign_key col, table, null: true/false` to specify the new column null-ness (`null: true` is default so you only need this for `null: false`). If you need to change the null-ness, it can/should be in another `alter_table` block. It may be the `index` line avoided the edge case bug- but you can also use `index: true` as a column option (ie, `add_foreign_key col, table, index: true`) to define the index.

> But somehow I am not finding this multiple migrations approach correct and esp running the migrations on different development envs already containing existing tables can become annoyance. So that makes me ask the question what is the right approach in Sequel to create a migration for altering an existing table by adding a not-null column with a default value?

This is not a multiple migrations approach. Many migrations require multiple steps. For example:

- Create a new FK column as nullable
- Find a row in the other table corresponding to some consistent field that is going to be the same for all environments, like a name/key. PK's will differ between environments so you can't reliably hard code a primary key in migrations.
- Update all rows to have this default value.
- Set the column as non-nullable. All rows are valid. Let the application enforce the default FK value.

Similarly, in the case of a state machine 'status' column:
- Create the `status` column as nullable
- Update all rows to the default state, which is correct *as of the migration* but can change in the future
- Set the column as non-nullable. The application from this moment forward enforces the default.

**Comment by jeremyevans** [comment]
You should set the default and NOT NULL when creating the column (you can also create an index at the same time):

```ruby
      alter_table(:my_table_name) do
        add_foreign_key :my_new_column, :other_table_name, :default=>0, :null=>false, :index=>true
      end
```

**Comment by jiggneshhgohel** [comment]
@rgalanakis, @jeremyevans Thanks a lot for your prompt response and esp Thanks to @rgalanakis for your elaborate reply and clarifications. Those clears up a lot of things for me.

Considering the elaborate details shared I think I was wrong somehow in my understanding about the `alter_table` API. But I will also say that I cannot completely blame myself for such partial understanding of API.  

If you *will* look at the documentation for `alter_table` API from a novice’s eyes (like me), then I think you should be able to understand what caused me to interpret that I should use `add_foreign_key`, `add_index`, `set_column_default`, etc separately.

Frankly speaking after this discussion what I feel is the documentation lacks  adequate/clear examples for the usage of various options available for the methods available inside the `alter_table` API. For e.g `add_foreign` key only mentions about first argument and constraint name.

Hoping that this feedback will be considered on a positive note.

Thanks.

**Comment by rgalanakis** [comment]
To preface this, I'm not a maintainer, but I've used Sequel across a half-dozen companies and dozen+ projects since 2014, when I started Ruby.

> Frankly speaking after this discussion what I feel is the documentation lacks adequate/clear examples for the usage of various options available for the methods available inside the alter_table API

That is probably true, even without looking at the docs to verify. For better and worse, Sequel conceptually leans pretty heavily on existing SQL knowledge as a baseline. It extensively documents things like its associations, which map foreign keys to the model system; but it doesn't, for example, automatically add indices to foreign keys, or other things many ORMs do because they think it's helpful. It is an 'SQL toolkit' (by far the best I've ever used across any language) which doesn't try to hide the database. Conceptually then for things like `alter_table` or `set_column_not_null`, the assumption is, learning about your RDBMS's `ALTER TABLE` goes hand in hand with Sequel's.

My opinion is this makes Sequel have a steeper initial learning curve, because it doesn't defacto need to hold your hand as closely, since the concepts aren't out of the Jeremy's head (who is not the original author) but are mostly found in SQL. But the learning curve is much more gentle after that, because it maps closely to SQL/your database's capabilities, and it's just a question of expressing a database idea in Sequel, rather than having to redesign your database idea so it works in an ORM.

I also can't verify this but I think the community skews to be more experienced, probably because we appreciate Sequel being such a good partner with the database, rather than competing with it.

All that said, I'm sure Jeremy would appreciate doc contributions to make them more friendly to beginners!

**Comment by jeremyevans** [comment]
> If you _will_ look at the documentation for `alter_table` API from a novice’s eyes (like me), then I think you should be able to understand what caused me to interpret that I should use `add_foreign_key`, `add_index`, `set_column_default`, etc separately.

There are definitely reasons you would want to issue the commands separately.  On some databases (older PostgreSQL IIRC), adding a column with a default value can lock the entire table while the whole table is rewritten, whereas adding a column without a default value does not.  `add_foreign_key` followed by `set_column_default` can perform better in such cases, but it does not set the default on the existing rows of the column, it only sets the default for future rows (you would need to update existing rows manually if you want).  This matches the behavior in SQL.

As @rgalanakis mentioned, Sequel's documentation assumes you know SQL fairly well in many cases.  This is great if do know SQL fairly well, but definitely suboptimal if you don't.  

There is no difference between using `add_index` versus the `:index` option, the behavior there is the same in both cases, since a separate query is always used to create the index.

> Frankly speaking after this discussion what I feel is the documentation lacks adequate/clear examples for the usage of various options available for the methods available inside the `alter_table` API. For e.g `add_foreign` key only mentions about first argument and constraint name.

`add_foreign_key` does state "See [CreateTableGenerator#column](http://sequel.jeremyevans.net/rdoc/classes/Sequel/Schema/CreateTableGenerator.html#method-i-column) for the available options", and all options are documented there. You are correct that is does not offer an example for your specific use case, or actually, an example of any of the documented options, and the same is true for most of the other methods documented in the same section.

I'm open to additional documentation examples.  However, I don't think it is feasible to document the recommended way to handle every possible case, especially since recommended handling is situation-dependent in many cases.

> Hoping that this feedback will be considered on a positive note.

Definitely.  Thank you for providing feedback.  If you have ideas for specific documentation improvements, submission of documentation pull requests is welcomed.

**Reply by jiggneshhgohel** [reply]
@jeremyevans to following 

>`add_foreign_key` does state "See [CreateTableGenerator#column](http://sequel.jeremyevans.net/rdoc/classes/Sequel/Schema/CreateTableGenerator.html#method-i-column) for the available options", and all options are documented there. You are correct that is does not offer an example for your specific use case, or actually, an example of any of the documented options, and the same is true for most of the other methods documented in the same section.

I would like to clarify here that when I referred to "documentation" in my comments I was referring to following doc https://github.com/jeremyevans/sequel/blob/master/doc/schema_modification.rdoc in repo but based on your reply you referred to the docs at http://sequel.jeremyevans.net/rdoc/classes/Sequel/Schema/AlterTableGenerator.html#method-i-add_foreign_key and latter does reference the supported options. So somewhere the discussion was not going on the same page.

From the day I have started working with Sequel migrations I am following the https://github.com/jeremyevans/sequel/blob/master/doc/schema_modification.rdoc because searching through Google for sequel migrations I have always got that document in top-search results (please refer attached screenshot) 

![1](https://github.com/jeremyevans/sequel/assets/10440841/78f14b19-c048-4d1c-a4a1-8abf4c3cd57e)

and if you will refer to the documentation at https://github.com/jeremyevans/sequel/blob/master/doc/schema_modification.rdoc#label-add_foreign_key you should be able to understand the source of my confusion.

So I would recommend that anybody ending up reading this discussion should refer to the docs pointed by @jeremyevans instead of what I was referring to.

@jeremyevans my apologies if this source of confusion caused any inconvenience but still multiple perspectives and facts were shared so definitely the discussion cannot be considered a waste.

Thanks.

**Comment by jiggneshhgohel** [comment]
@jeremyevans Thanks for your elaborate reply.

>As [@rgalanakis](https://github.com/rgalanakis) mentioned, Sequel's documentation assumes you know SQL fairly well in many cases. This is great if do know SQL fairly well, but definitely suboptimal if you don't.

That’s fair. 

But I would like to share here my subjective feelings regarding the need for documentation to be clear/comprehensive in its approach irrespective of that user be experienced or novice with Sequel. Below are my thoughts:


For developers, whose mindset is habitual to using ORMs for working with databases, the assumption 

>Sequel's documentation assumes you know SQL fairly well in many cases.

can turn out contrary. And a reason behind my saying so is following:

Majority of the popular frameworks facilitating working with databases through ORMs is in vogue and from past so many years working in the industry I constantly feel that ORMs have contributed towards detaching  developers from working with raw SQL. And to extend my feeling to the extreme I would say that if a survey be done with majority of the programmers initiating their career into programming through frameworks like Rails and if they are asked how do you work with database the answer should be ActiveRecord ORM. 

And no wonder that who purely relies on working with ORMs contributes towards more performance issues from DB-front  in the application because majority of them would not be even aware about the queries that should be firing at DB-level.

From our discussion in previous comments what I have understood is that Sequel is not inclined towards being an ORM in literal terms but a high-level interface towards working with DB and still remaining in touch with the fundamentals of SQL and facilitating the performance gains of working with raw SQL. 

For a developer switching from his existing  language domain (like Java, Python etc) to Ruby language domain, because of the common psychology of people following the trends, Rails is the framework on which such developers end-up. 

And Rails by default supports ActiveRecord ORM. So if such developers who have worked with Rails for quite sometime and for any reason they need to switch to or explore another ORM, then the preferred option is found to be Sequel and recently a still high-level interface is being developed by rom-rb community with the focus being resolving the performance issues involved with ActiveRecord. 

Now for working effectively with with rom-rb Sequel’s knowledge is an expectation. 

Now if such a developer (and that includes me as well) ends up on Sequel’s documentation (with a novice-mindset to Sequel), obviously they would look at it through a biased mindset wherein they should be assuming that they are switching to Sequel in lieu of another ORM, like ActiveRecord they are familiar with, and hence their learning curve should be fairly smooth but the Sequel’s community will only know that that assumption is wrong. 

And the more experienced such developer would be and who has became fairly familiar with fundamentals of working with Sequel, the more he would be targeting  directly the API documentation for his advanced needs. But since the API documentation is not comprehensive or directly indicative at various places (for e.g the `add_foreign_key` documentation) he should end up in the same situation like I shared in the foremost comment in this discussion.  

Please note that what all I have written in this comment or in previous comments has no intention of getting into any sort of debates and argumentation. I have simply and frankly shared my feelings about the need for improving Sequel’s documentation so that a mere skimmer also can become aware about what all a particular API has to offer. 

Lastly I would like to say that: It’s our common experience that without comprehensive and clear manuals we cannot take full advantage of what a great utility thing or a mighty gadget has to offer and because of lack of proper usage instructions be whatever potential a thing has, it cannot be leveraged. Similar thought I have about the open-source tools APIs. 

But in Sequel case a great effort has been put into drafting documentation as is evident from the website but there still is  need is for making it clear at various places by elaborating with more examples for various options an API has to offer or adding at-least references to all options it provides and where the examples for those options usage can be found. 

Finally Thank you for taking the time to respond to my queries with as much elaboration as possible. It has definitely helped in looking at Sequel with right perspective. 

And I highly appreciate your prompt responses to the queries for the open-source tools you have offered to the community. I can recollect the same feeling when I posted queries about Roda’s usage. You simply are awesome!

Thanks.

**Comment by jeremyevans** [comment]
> For developers, whose mindset is habitual to using ORMs for working with databases, the assumption
> 
> > Sequel's documentation assumes you know SQL fairly well in many cases.
> 
> can turn out contrary.

Certainly.  But documentation cannot optimize for all possible audiences.  If you optimize the documentation for people who don't know SQL, you end up making the documentation worse for users who know SQL, and vice-versa.  Sequel's documentation deliberate chooses to optimize documentation for users that know SQL.  I realize that makes the documentation  suboptimal for users that do not know SQL, but I would rather the documentation be more useful for users who know SQL.

> Majority of the popular frameworks facilitating working with databases through ORMs is in vogue and from past so many years working in the industry I constantly feel that ORMs have contributed towards detaching developers from working with raw SQL.

I would say that is not a good thing.  While Sequel does not encourage the use of raw SQL fragments (as that often can lead to SQL injections), I would say it encourages the full use of SQL concepts.

> And to extend my feeling to the extreme I would say that if a survey be done with majority of the programmers initiating their career into programming through frameworks like Rails and if they are asked how do you work with database the answer should be ActiveRecord ORM.

I would assume so as well.  However, I would also assume that Rails developers who know and understand SQL are probably far more productive than Rails developers who do not.

> And no wonder that who purely relies on working with ORMs contributes towards more performance issues from DB-front in the application because majority of them would not be even aware about the queries that should be firing at DB-level.

Which I think is a problem.  IMO, developers should be aware of what queries are being used.  ORMs should be used to make things easier for the user to interact with a database, not as a black box.

> From our discussion in previous comments what I have understood is that Sequel is not inclined towards being an ORM in literal terms but a high-level interface towards working with DB and still remaining in touch with the fundamentals of SQL and facilitating the performance gains of working with raw SQL.

Correct.  Sequel itself is not an ORM, it is a database access layer.  Sequel::Model is an ORM that is built on top of that database access layer.  ROM is another ORM built on top of that database access layer.

> For a developer switching from his existing language domain (like Java, Python etc) to Ruby language domain, because of the common psychology of people following the trends, Rails is the framework on which such developers end-up.
> 
> And Rails by default supports ActiveRecord ORM. So if such developers who have worked with Rails for quite sometime and for any reason they need to switch to or explore another ORM, then the preferred option is found to be Sequel and recently a still high-level interface is being developed by rom-rb community with the focus being resolving the performance issues involved with ActiveRecord.
> 
> Now for working effectively with with rom-rb Sequel’s knowledge is an expectation.
> 
> Now if such a developer (and that includes me as well) ends up on Sequel’s documentation (with a novice-mindset to Sequel), obviously they would look at it through a biased mindset wherein they should be assuming that they are switching to Sequel in lieu of another ORM, like ActiveRecord they are familiar with, and hence their learning curve should be fairly smooth but the Sequel’s community will only know that that assumption is wrong.
> 
> And the more experienced such developer would be and who has became fairly familiar with fundamentals of working with Sequel, the more he would be targeting directly the API documentation for his advanced needs. But since the API documentation is not comprehensive or directly indicative at various places (for e.g the `add_foreign_key` documentation) he should end up in the same situation like I shared in the foremost comment in this discussion.

I suppose that may be true for some experienced developers, but certainly not for all. You should not project that other users will generally have the same experience you had, because they come from different backgrounds, and especially not in the case where the hypothetical user would have more knowledge than you have.  A experienced SQL user would likely know the difference between separate queries to add a column and set a default, versus a single query to add a column with a default.  And the documentation for the method does link to the list of comprehensive options, even if no example of those options is given.

> Please note that what all I have written in this comment or in previous comments has no intention of getting into any sort of debates and argumentation. I have simply and frankly shared my feelings about the need for improving Sequel’s documentation so that a mere skimmer also can become aware about what all a particular API has to offer.

That's fine.  I also am not attempting to debate, merely offering my perspective on the situation.

> Lastly I would like to say that: It’s our common experience that without comprehensive and clear manuals we cannot take full advantage of what a great utility thing or a mighty gadget has to offer and because of lack of proper usage instructions be whatever potential a thing has, it cannot be leveraged. Similar thought I have about the open-source tools APIs.
> 
> But in Sequel case a great effort has been put into drafting documentation as is evident from the website but there still is need is for making it clear at various places by elaborating with more examples for various options an API has to offer or adding at-least references to all options it provides and where the examples for those options usage can be found.

As pointed out earlier, the documentation in question does reference all options.  It doesn't specifically give an example of those options.  Again, I'm open to additional examples.

> Finally Thank you for taking the time to respond to my queries with as much elaboration as possible. It has definitely helped in looking at Sequel with right perspective.
> 
> And I highly appreciate your prompt responses to the queries for the open-source tools you have offered to the community. I can recollect the same feeling when I posted queries about Roda’s usage. You simply are awesome!

Thank you very much!
